include/garden/system/timer.hpp

```diff
@@ -0,0 +1,145 @@
+/**
+ * \file timer.hpp
+ * \author Sahak Grigoryan <sahakgrigoryan.am@gmail.com>
+ * 
+ * \brief Class definition: TimerSystem
+ */
+
+#pragma once
+
+//--------------------------------------------------------------------------
+// includes
+
+// C/C++ standard headers
+#include <chrono>
+#include <unordered_set>
+
+// third-party library headers
+#include "ecsm.hpp"
+
+// project-specific headers
+
+//--------------------------------------------------------------------------
+// forward declarations
+namespace ecsm
+{
+template <class, bool>
+class LinearPool;
+} // namespace ecsm
+
+//--------------------------------------------------------------------------
+// class definition and member declarations
+namespace garden
+{
+
+using ecsm::IdHash;
+using ecsm::ID;
+using ecsm::Singleton;
+using ecsm::System;
+using ecsm::View;
+
+/**
+ * \class TimerSystem
+ * \brief Manages timer-based event handlers.
+ */
+class TimerSystem final : public System, public Singleton<TimerSystem>
+{
+public:
+    using ClockType     = std::chrono::steady_clock;
+    using TimeUnit      = std::chrono::milliseconds;
+    using TimePoint     = std::chrono::time_point<ClockType, TimeUnit>;
+    using TimerCallback = std::function<bool(void)>;
+
+    /**
+     * \brief Timer metadata structure.
+     */
+    struct TimerHandler {
+        TimePoint     initTime; //< the time point of timer registration and refresh (\see registerTimer())
+        TimeUnit      duration; //< the time duration (ms) that has to elapse before event handling
+        TimerCallback callback; //< the event callback after time elapses
+        bool          isOneShot; //< specifies if timer to be executed once (\see registerOneShotTimer()) or periodically (\see registerTimer())
+    };
+
+    /// constructor
+    TimerSystem(void);
+
+    /// destructor
+    virtual ~TimerSystem(void) override final;
+
+    /**
+     * \brief Runs and manages all timer handlers; "Timer" ordered event handler.
+     */
+    void runTimers(void);
```

Comment:
This should be private, because it's subscribed to the "Timer" event which managed by the Manager.